src/goal.rs

```diff
@@ -1194,11 +1329,46 @@ pub fn prove(mut goal: Goal) -> (Outcome, ProofState) {
       }
       return (Outcome::Unknown, state);
     }
-    goal.case_split(&mut state);
-    if CONFIG.verbose {
-      println!("{}", "Case splitting and continuing...".purple());
+    if let Some(scrutinee) = goal.next_scrutinee() {
+      if CONFIG.verbose {
+        println!("{}", "Case splitting and continuing...".purple());
+      }
+      goal.case_split(scrutinee, &mut state);
+    } else {
+      if CONFIG.verbose {
+        println!("{}", "Cannot case split: no blocking variables found".red());
+        for remaining_goal in &state.goals {
+          println!("{} {}", "Remaining case".yellow(), remaining_goal.name);
+        }
+      }
+      return (Outcome::Invalid, state);
     }
   }
   // All goals have been discharged, so the conjecture is valid:
   (Outcome::Valid, state)
 }
+
+fn cartesian_product_helper<T: Clone>(
+  vector: &[Vec<T>],
+  index: usize,
+  current_combination: Vec<T>,
+  result: &mut Vec<Vec<T>>,
+) {
+  if index == vector.len() {
+    result.push(current_combination);
+  } else {
+    for elem in &vector[index] {
+      let mut new_combination = current_combination.clone();
+      new_combination.push(elem.clone());
+      cartesian_product_helper(vector, index + 1, new_combination, result);
+    }
+  }
+}
+
+/// Given a vector of vectors, generates the "cartesian product" of all the vectors.
+/// TODO: figure out how to use multi_cartesian_product from itertools instead of writing our own
+fn cartesian_product<T: Clone>(vector: &[Vec<T>]) -> Vec<Vec<T>> {
```

Comment:
We already have an efficient Cartesian product computation inside of `collect_expressions` in `egraph.rs`, although it's not factored out into a separate function. Can you try to reuse that one if possible? (I.e. refactor that code into a separate function and use it in both places)

Reply:
I don't think `add_grounding` has any Cartesian product-like code anymore. The implementation with canonical forms is very different (it just takes all the previous instantiations and replaces a given parent var with a given child var wherever it can).